desc/efit.py

```diff
@@ -0,0 +1,374 @@
+                g["nbbbs"] = len(g["rbbbs"])
+                plt.close()
+
+        ns = len(g["q"])
+        chiN = np.linspace(0, 1, ns)
+        chi_scale = g["chi_boundary"] - g["chi_axis"]
+        # chi is the poloidal flux
+        # psi is the toroidal flux
+
+        # iota = dchi / dpsi = dchi / dchiN * dchiN / dpsi = dchiN / dpsi * (chi_scale)
+        # dpsi = 1/iota dchi = q dchiN
+        # psi = integral(q dchiN) * (chi_scale)
+
+        # q * (chi(1) - chi(0)) as function of chiN
+        q_spline = CubicSpline(chiN, g["q"] * chi_scale)
+        # this gives psi(chiN)
+        psi_spline = q_spline.antiderivative()
+
+        psi = psi_spline(chiN)
+        Psi_b = psi[-1]  # boundary flux
+
+        pressure_psi = SplineProfile(g["pres"], psi)
+        iota_psi = SplineProfile(1 / g["q"], psi)
+
+        # Defining splines and recalculating important quantities if
+        # bdry_dist less than 1. Also works for bdry_dist equal to 1.
+        # For bdry_dist equal 1, all quantities *_truncated are the
+        # same as quantities without _truncated
+        chiN_truncated = np.linspace(0, bdry_dist, ns)
+        q_spline_truncated = CubicSpline(chiN_truncated, q_spline(chiN_truncated))
+
+        psi_spline_truncated = q_spline_truncated.antiderivative()
+        psi_truncated = psi_spline_truncated(chiN_truncated)
+        Psi_b_truncated = psi_truncated[-1]  # boundary flux
+        psiN_truncated = psi_truncated / Psi_b_truncated
+        rho_truncated = np.sqrt(psiN_truncated)
+
+        pressure = SplineProfile(pressure_psi(psi_truncated), rho_truncated)
+        iota = SplineProfile(iota_psi(psi_truncated), rho_truncated)
+
+        nbdry = g["nbbbs"]
+        rbdry = g["rbbbs"]
+        zbdry = g["zbbbs"]
+        theta = np.linspace(0, 2 * np.pi, nbdry)  # rbbbs includes endpoints
+
+        grid = LinearGrid(theta=theta)
+        basis = DoubleFourierSeries(M=M, N=N, sym=False)
+        transform = Transform(grid, basis, build_pinv=True)
+        Rb_lmn = transform.fit(rbdry)
+        Zb_lmn = transform.fit(zbdry)
+
+        surf = FourierRZToroidalSurface(
+            R_lmn=Rb_lmn,
+            Z_lmn=Zb_lmn,
+            modes_R=basis.modes[:, 1:],
+            modes_Z=basis.modes[:, 1:],
+        )
+
+        inputs = {}
+
+        inputs["pressure"] = pressure
+        inputs["iota"] = iota
+        inputs["sym"] = False
+        inputs["surface"] = surf
+        inputs["NFP"] = int(1)
+        inputs["Psi"] = Psi_b
+        inputs["L"] = L
+        inputs["M"] = M
+        inputs["N"] = N
+
+        eq = Equilibrium(**inputs)
+
+        return eq
+
+    @classmethod
+    def efit_to_desc_input(  # noqa: C901 - fxn too complex
+        cls,
+        eq,
+        outfile,
+        objective="force",
+        optimizer="lsq-exact",
+        ftol=1e-2,
+        xtol=1e-2,
+        gtol=1e-6,
+        maxiter=100,
+        threshold=1e-8,
+    ):
+        """Generate a DESC input file from a DESC equilibrium object.
```

Comment:
this docstring sohuld mention efit?? what is the purpose of the function?

Reply:
You are right. This function is more general purpose than an input file writer for EFIT files. Right now, I don't know how to move it into another script.